docs/proposals/perf.md

```diff
@@ -0,0 +1,381 @@
+
+Performance Test Framework features include:
+
+- A comprehensive test runner.
+- Built-in support for testing asynchronicity.
+- Modular and easy to customize.
+- Logging and Reporting.
+- Scalable to add more features.
+- Built-in support of command line interface.
+- ...
+
+### Performance Test Metrics Tools
+* [Prometheus](https://github.com/prometheus/prometheus)
+* [Grafana](https://github.com/grafana/grafana)
+
+### Performance Test Scenarios
+
+#### 1. Edge Nodes join in K8S Cluster
+<img src="../images/perf/perf-edgenodes-join-cluster.png">
+
+Test Cases:
+* Measure Edge Nodes join in K8S Cluster startup time.
+
+  Different numbers of Edge Nodes need be tested.
+
+  Edge Nodes numbers are one of `[1, 10, 20, 50, 100, 200...]`.
+
+  This test case ends with all Edge Nodes are in `Ready` status.
+
+* Measure CPU and Memory Usage of KubeEdge Cloud Part.
+
+  Edge Nodes numbers are one of `[1, 10, 20, 50, 100, 200...]`.
+
+* Measure CPU and Memory Usage of KubeEdge Edge Part.
+
+  Edge Nodes numbers are one of `[1, 10, 20, 50, 100, 200...]`.
+
+#### 2. Create Devices from Cloud
+<img src="../images/perf/perf-create-device.png">
+
+This scenario is expected to measure the northbound API of KubeEdge.
+
+Test Cases:
+* Measure the latency between K8S Master and KubeEdge Cloud Part.
+* Measure the throughput between K8S Master and KubeEdge Cloud Part.
+* Measure CPU and Memory Usage of KubeEdge Cloud Part.
+
+#### 3. Report Device Status to Edge
+<img src="../images/perf/perf-report-devicestatus.png">
+
+This scenario is expected to measure the southbound API of KubeEdge.
+
+Test Cases:
+* Measure the latency between KubeEdge Edge Part and device.
+
+  Device numbers per Edge Node are one of `[1, 10, 20, 50, 100, 200...]`.
+
+* Measure the throughput between KubeEdge Edge Part and device.
+
+  Device numbers per Edge Node are one of `[1, 10, 20, 50, 100, 200...]`.
+
+* Measure CPU and Memory Usage of KubeEdge Edge Part.
+
+  Device numbers per Edge Node are one of `[1, 10, 20, 50, 100, 200...]`.
+
+As the result of the latency and throughput with different device numbers,
+we can evaluate scalability of devices for KubeEdge Edge Part.
+Measure how many devices can be handled per Edge Node.
+<img src="../images/perf/perf-multi-devices.png">
+
+Different protocols are considered to test between KubeEdge Edge Part and devices.
+E.g. Bluetooth, MQTT, ZigBee, BACnet and Modbus and so on.
+Currenly less than 20ms latency can be accepted in Edge IoT scenario.
+Two kinds of test cases can be adopted: emulators of different devices and actual devices.
+
+#### 4. Application Deployment from Cloud to Edge
+<img src="../images/perf/perf-app-deploy.png">
+
+This scenario is expected to measure the performance of KubeEdge from Cloud to Edge.
+The docker image download latency is not included in this scenario.
+In the following test cases, we need to make sure that docker images have already been in the Edge Node.
```

Comment:
have already been in the Edge Node -> are already downloaded on the edge node.

Reply:
Thanks. will update later